src/cts/src/HTreeBuilder.cpp

```diff
@@ -224,6 +225,456 @@ void HTreeBuilder::initSinkRegion()
+                                   unsigned i,
+                                   unsigned par)
+{
+  unsigned idx = 0;
+  for (; idx < topology.getBranchingPointSize(); ++idx) {
+    unsigned k = topology.getBranchingPointParentIdx(idx);
+    if (idx != i && k == par) {
+      return idx;
+    }
+  }
+  return i;
+}
+
+void scalePosition(Point<double>& loc,
+                   const Point<double>& parLoc,
+                   double leng,
+                   double scale)
+{
+  double px = parLoc.getX();
+  double py = parLoc.getY();
+  double ax = loc.getX();
+  double ay = loc.getY();
+
+  double d = loc.computeDist(parLoc);
+  double x, y;
+  if (d > 0) {  // yy8
+    double delta = d * scale;
+    double dx = ax - px;
+    double dy = ay - py;
+    dx += (dx > 0) ? delta : -delta;
+    dy += (dy > 0) ? -delta : delta;
+    double scale = leng / d;
+    x = px + dx * scale;
+    y = py + dy * scale;
+  } else {
+    x = px + leng / 2;
+    y = py + leng / 2;
+  }
+  loc.setX(x);
+  loc.setY(y);
+}
+
+void setSiblingPosition(const Point<double>& a,
+                        Point<double>& b,
+                        const Point<double>& parLoc)
+{
+  double px = parLoc.getX();
+  double py = parLoc.getY();
+  double ax = a.getX();
+  double ay = a.getY();
+  double bx = 2 * px - ax;
+  double by = 2 * py - ay;
+  b.setX(bx);
+  b.setY(by);
+}
+
+/* xxx Balance the two branches on the very top level */
+void adjustToplevelTopology(Point<double>& a,
+                            Point<double>& b,
+                            const Point<double>& parLoc)
+{
+  double da = a.computeDist(parLoc);
+  double db = b.computeDist(parLoc);
+  if (da < db) {
+    setSiblingPosition(a, b, parLoc);
+  } else if (db < da) {
+    setSiblingPosition(b, a, parLoc);
+  }
+}
+
+bool moveOnBlockBounardy(const Point<double>& pt,
+                         Point<double>& qt,
+                         double& x1,
+                         double& y1,
+                         double& x2,
+                         double& y2)
+{
+  double px = pt.getX();
+  double py = pt.getY();
+  double qx = qt.getX();
+  double qy = qt.getY();
+
+  // assert ( (!px>=x1 && px <x2 && py>=y1 && py <= y2) ) ;
+  // assert ( qx>=x1 && qx <x2 && qy>=y1 && qy <= y2 ) ;
+
+  double dx = px - qx;
+  double dy = py - qy;
+
+  std::vector<Point<double>> points;
+  if (dx == 0 || dy == 0) {  // vertical or horizontal
+    points.emplace_back(qx, y1);
+    points.emplace_back(qx, y2);
+    points.emplace_back(x1, qy);
+    points.emplace_back(x2, qy);
+  } else {
+    double m = dy / dx;
+    points.emplace_back(x1, m * (x1 - qx) + qy);  // y = m*(x-qx) + qy
+    points.emplace_back(x2, m * (x2 - qx) + qy);
+    points.emplace_back((y1 - qy) / m + qx, y1);  // x = (y-qy)/m + qx
+    points.emplace_back((y2 - qy) / m + qx, y2);
+  }
+  double d1 = pt.computeDist(qt);
+  for (Point<double> u : points) {
+    double d2 = u.computeDist(pt) + u.computeDist(qt);
+    if (abs(d1 - d2) < d1 / 100000) {
```

Comment:
What is this magic number?

Reply:
I see some issue with this routine beyond the magic number comparison.   Please allow for this version to be checked in first.  I'll make improvements later.

Reply:
Added TODO comment